src/python/pants/backend/typescript/dependency_inference/rules.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import itertools
+import os.path
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.backend.javascript import package_json
+from pants.backend.javascript.dependency_inference.rules import (
+    InferNodePackageDependenciesRequest,
+    NodePackageCandidateMap,
+    RequestNodePackagesCandidateMap,
+)
+from pants.backend.javascript.package_json import (
+    OwningNodePackage,
+    OwningNodePackageRequest,
+    PackageJsonSourceField,
+)
+from pants.backend.javascript.subsystems.nodejs_infer import NodeJSInfer
+from pants.backend.typescript.target_types import (
+    TS_FILE_EXTENSIONS,
+    TypeScriptDependenciesField,
+    TypeScriptSourceField,
+)
+from pants.build_graph.address import Address
+from pants.engine.addresses import Addresses
+from pants.engine.internals.graph import Owners, OwnersRequest
+from pants.engine.internals.native_dep_inference import NativeParsedJavascriptDependencies
+from pants.engine.internals.native_engine import InferenceMetadata, NativeDependenciesRequest
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.engine.target import (
+    FieldSet,
+    HydratedSources,
+    HydrateSourcesRequest,
+    InferDependenciesRequest,
+    InferredDependencies,
+    Targets,
+)
+from pants.engine.unions import UnionRule
+from pants.util.ordered_set import FrozenOrderedSet
+
+
+@dataclass(frozen=True)
+class TypeScriptSourceInferenceFieldSet(FieldSet):
+    required_fields = (TypeScriptSourceField, TypeScriptDependenciesField)
+
+    source: TypeScriptSourceField
+    dependencies: TypeScriptDependenciesField
+
+
+@dataclass(frozen=True)
+class TypeScriptFileImportPath:
+    """Path to a file that is imported in TypeScript code."""
+
+    path: str
+
+
+class InferTypeScriptDependenciesRequest(InferDependenciesRequest):
+    infer_from = TypeScriptSourceInferenceFieldSet
+
+
+async def _prepare_inference_metadata(address: Address) -> InferenceMetadata:
+    owning_pkg = await Get(OwningNodePackage, OwningNodePackageRequest(address))
+    if not owning_pkg.target:
+        return InferenceMetadata.javascript(address.spec_path, {})
+    return await Get(
+        InferenceMetadata, PackageJsonSourceField, owning_pkg.target[PackageJsonSourceField]
+    )
+
+
+@rule
+async def get_file_imports_targets(import_path: TypeScriptFileImportPath) -> Targets:
+    """Get address to build targets representing a file import discovered with dependency inference.
+
+    For now, we only support .ts files; in the future, may need to iterate through all possible file
```

Comment:
Is there a good reason to not also support `.js` immediately?

Reply:
I've tried to do that, and it resulted in two extra rules and a few `MultiGet` calls to iterate through all supported extensions, so I wanted to keep it simpler first, one PR = one "feature" / functionality piece. It's also overwhelming for me to get all the bits in place as I am still exploring the TypeScript world :)

However, even with rules, when attempting to get a target that owns a JS module, I get an error, so I must be missing something. In a next PR, I'll show what I do and maybe get some help :)

Reply:
It may be worth making an `AllImportableTypescriptTargets` singleton param that fetches _all_ of the ts/js/etc targets and indexes them in a dict (`DjangoApps` in `backend/python/framework/django/detect_apps` is a good example of what I mean). Then you could have a mapping from module name to the target name. Given that `AllImportableTypesciptTargets` would only be computed once, it should be reasonably cheap and will probably be easer to implement than a `MultiGet` would be.

Reply:
Thanks for this, I'll need to explore. Would you be comfortable doing this in a follow-up to add support for JS dependencies or would rather want to see this already in this PR? :)

Reply:
I would prefer any decision on this is defered to after more building blocks are in place (see https://github.com/pantsbuild/pants/pull/21176).